[Java] Adds support for DNNL, OpenVINO, TensorRT shared providers and refactors the CUDA shared provider loader

Craigacp opened the pull request:
**Description**:

Refactors the native library loading in Java to allow CUDA to be loaded on demand, fixing #7044. Then expands the shared provider library loading to DNNL, OpenVINO, TensorRT, fixing #6553.

Added a flag to the native library loading to allow users to supply a directory which contains all the native libraries, fixing #8003. This is also the only way to make the shared library providers load from a different place than the jar, as the individual library path specification conflicts with the way that the ONNX Runtime native code loads the shared library providers.

I also slightly refactored the Java cmake bits, and added the `--console=plain` flag to the gradle executions to stop gradle writing over cmake's output.

**Motivation and Context**
- Why is this change required? What problem does it solve? Re-enables DNNL, OpenVINO and TensorRT in Java by allowing them to be packaged in the jar and dynamically loaded in the same way CUDA is.
- If it fixes an open issue, please link to the issue here. Fixes #6553. Fixes #7044. Fixes #8003.

Craigacp commented:
One remaining todo with this PR is to enable OpenVINO, DNNL and TensorRT in the Java testing pipeline (or enable Java in the relevant testing pipelines there). Is there a preference which way round that happens?

Craigacp commented:
@RyanUnderhill @yuslepukhin

RyanUnderhill commented:
> One remaining todo with this PR is to enable OpenVINO, DNNL and TensorRT in the Java testing pipeline (or enable Java in the relevant testing pipelines there). Is there a preference which way round that happens?

I think it's easier to add it to the provider pipelines (since provider availability can sometimes be mutually exclusive) vs trying to make them all work in one Java pipeline. For example, this isn't a shared provider yet, but I don't expect to be able to have CUDA and ROCM at the same time.

Thank you for cleaning up my original shared library stuff, this definitely makes it more efficient.

Craigacp commented:
Ok, I've added Java tests for TensorRT, OpenVINO and DNNL, and I added `--build_java` to the DNNL and OpenVINO pipelines (it was already present on the TensorRT one, though until this PR that wouldn't have done anything different to the CPU version of the Java code).

I'm not sure I've got the pipeline changes right, and I can't easily run them, so I'd appreciate it if someone could look over them and kick off the pipelines.

snnn commented:
/azp run Linux CPU CI Pipeline,  Linux CPU Minimal Build E2E CI Pipeline,  Linux CPU x64 NoContribops CI Pipeline,  Linux GPU CI Pipeline

snnn commented:
/azp run Linux GPU TensorRT CI Pipeline,  Linux Nuphar CI Pipeline,  Linux OpenVINO CI Pipeline,  MacOS CI Pipeline,  MacOS NoContribops CI Pipeline,  Windows CPU CI Pipeline,  Windows GPU CI Pipeline,  Windows GPU TensorRT CI Pipeline

snnn commented:
/azp run Windows WebAssembly CI Pipeline,  orttraining-amd-gpu-ci-pipeline,  orttraining-linux-ci-pipeline,  orttraining-linux-gpu-ci-pipeline,  orttraining-ortmodule-distributed

Craigacp commented:
Looks like the TensorRT test was skipped, I'll double check how it's passing in the flag. OpenVINO failed because it was looking for Java 8. I'm not sure what's setting `JAVA_HOME` in that docker file. Any ideas?

snnn commented:
> Looks like the TensorRT test was skipped by gradle, I'll double check how it's passing in the flag. OpenVINO failed because it was looking for Java 8. I'm not sure what's setting `JAVA_HOME` in that docker file. Any ideas?

The openvino build logs says: 
```
update-alternatives: using /usr/lib/jvm/java-11-openjdk-amd64/bin/java to provide /usr/bin/java (java) in auto mode
```

So it should be able to find java in PATH. Is setting JAVA_HOME still needed?

snnn commented:
```
ERROR: JAVA_HOME is set to an invalid directory: /usr/lib/jvm/java-8-openjdk-amd64

```
Oh, that's strange.

snnn commented:
I think you need to update or remove this: 

https://github.com/microsoft/onnxruntime/blob/master/tools/ci_build/github/linux/run_build.sh#L64

snnn commented:
@Craigacp I think you need to update this file as well:

https://github.com/microsoft/onnxruntime/blob/master/cmake/onnxruntime_java_unittests.cmake

Craigacp commented:
> @Craigacp I think you need to update this file as well:
> 
> https://github.com/microsoft/onnxruntime/blob/master/cmake/onnxruntime_java_unittests.cmake
> 
> You may use this `GRADLE_ARGS` flag:
> 
> https://github.com/microsoft/onnxruntime/blob/master/cmake/onnxruntime_java.cmake#L162

I've already got the ep flags in there as `GRADLE_TEST_EP_FLAGS`. Weirdly it seems to have run the CUDA test in the TensorRT pipeline, but not the TensorRT test. I've updated the Java test cmake file to make it print the flags, hopefully that will give more insight.

I also remove the `JAVA_HOME` setting from `run_build.sh`. Could you kick off the TensorRT, OpenVINO and DNNL pipelines to check it out?

Craigacp commented:
I've updated this PR to remove the merge conflicts, could someone start the CI pipelines?

Craigacp commented:
Could someone please start the CI pipelines? @snnn @RyanUnderhill

edgchen1 commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux CPU x64 NoContribops CI Pipeline, Linux GPU CI Pipeline, Windows WebAssembly CI Pipeline, orttraining-amd-gpu-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed

edgchen1 commented:
/azp run Linux GPU TensorRT CI Pipeline, Linux Nuphar CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, MacOS NoContribops CI Pipeline, Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline

Craigacp commented:
Looks like I put the argument outside the quote when it should have been inside the quote for OpenVINO. Could you restart that please?

snnn commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux CPU x64 NoContribops CI Pipeline, Linux GPU CI Pipeline, Windows WebAssembly CI Pipeline, orttraining-amd-gpu-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed

snnn commented:
/azp run Linux GPU TensorRT CI Pipeline, Linux Nuphar CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, MacOS NoContribops CI Pipeline, Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline

snnn commented:
"Failed to load native library 'libnative-platform.so' for Linux amd64." 

Do you know why it happens?

Craigacp commented:
Looks like that can be caused by the gradle home directory not being writable. I'm not too familiar with how the docker containers are setup, any idea where we can get some writable directories on them?

Craigacp commented:
The GPU TensorRT test passed in USE_CUDA rather than USE_TENSORRT - `7: Test command: /usr/local/gradle/bin/gradle "cmakeCheck" "-DcmakeBuildDir=/build/Release" "-DUSE_CUDA=1"`. Not sure how that happened.

snnn commented:
> gradle home directory
You mean /usr/local/gradle ? It is not writable.

snnn commented:
TRT's log says:
"-- GRADLE_ARGS: --console=plain;cmakeBuild;-DcmakeBuildDir=/build/Release;-DUSE_CUDA=1 -DUSE_TENSORRT=1"
I feel it should be
"-- GRADLE_ARGS: --console=plain;cmakeBuild;-DcmakeBuildDir=/build/Release;-DUSE_CUDA=1;-DUSE_TENSORRT=1"
It missed a ";"

"-DUSE_CUDA=1" and "-DUSE_TENSORRT=1" are two options, not one.

Craigacp commented:
> > gradle home directory
> 
> You mean /usr/local/gradle ? It is not writable.

No, usually it's `/home/<username>/.gradle`. It wants to store the lock file that controls the gradle daemon, and cache things like the protobuf jar there.

Craigacp commented:
I pushed out an update for the TensorRT thing, I might not have understood how CMake's argument lists work and hopefully this will be better.

I've had a look through the dockerfiles for OpenVINO and Ubuntu and I'm not sure what makes the OpenVINO one different from the regular Ubuntu one which I think is used for the Java builds which work. Any ideas where to look?

snnn commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux CPU x64 NoContribops CI Pipeline, Linux GPU CI Pipeline, Windows WebAssembly CI Pipeline, orttraining-amd-gpu-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed

snnn commented:
/azp run Linux GPU TensorRT CI Pipeline, Linux Nuphar CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, MacOS NoContribops CI Pipeline, Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline

Craigacp commented:
I fixed the test passthrough and checked it against DNNL which is the easiest for me to run. I'd missed that the build.gradle needs to pass through the system properties to the test runner, and that `onnxruntime_java_unittests.cmake` is only used on Windows.

I still don't understand why the OpenVINO test is failing. Is there someone on your side who understands how that's setup that could have a look at the error? They might understand why the filesystem doesn't appear to be writeable.

snnn commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux CPU x64 NoContribops CI Pipeline, Linux GPU CI Pipeline, Windows WebAssembly CI Pipeline, orttraining-amd-gpu-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed

snnn commented:
/azp run Linux GPU TensorRT CI Pipeline, Linux Nuphar CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, MacOS NoContribops CI Pipeline, Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline

Craigacp commented:
Well at least TensorRT is happy now - `7: InferenceTest > testTensorRT() PASSED`.

snnn commented:
I will look the OpenVINO  failure.  Please give me a few days.

Craigacp commented:
Thanks. If it's going to be a complicated fix then I'll turn off the OpenVINO test in Java so we can get this merged. I'm planning to update the Java provider interfaces so you can supply the various parameter structs to the CUDA provider etc and this change blocks most of them as it modifies the same files, so I'd appreciate getting this merged in early next week if we can.

snnn commented:
Good suggestion. Let's merge this PR first.

snnn commented:
Please help turn off the OpenVINO test for now.

Craigacp commented:
Done.

snnn commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux CPU x64 NoContribops CI Pipeline, Linux GPU CI Pipeline, Windows WebAssembly CI Pipeline, orttraining-amd-gpu-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed

snnn commented:
/azp run Linux GPU TensorRT CI Pipeline, Linux Nuphar CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, MacOS NoContribops CI Pipeline, Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline

snnn commented:
Linux GPU CI Pipeline failure is due to an unknown reason that I'm still investigating.

snnn commented on tools/ci_build/github/linux/docker/Dockerfile.ubuntu_openvino:
```diff
@@ -47,4 +48,4 @@ ARG BUILD_USER=onnxruntimedev
 WORKDIR /home/$BUILD_USER
 RUN adduser --gecos 'onnxruntime Build User' --disabled-password $BUILD_USER --uid $BUILD_UID
 RUN adduser $BUILD_USER video
```
I think this is why the openvino build failed. It created two users! The one used for running CI build doesn't have the necessary permissions. And the one has perm wasn't used for doing anything.

Craigacp replied:
Ah ok. Well that sounds like a problem. To re-enable the java tests I think all that's necessary is to add `--build_java` to the end of the docker run command in linux-openvino-ci-pipeline.yml, inside the double quotes.
